packages/playground/sync/src/sql.ts

```diff
@@ -110,60 +116,56 @@ export async function recordSQLQueries(
 		}
 
 		if (
-			message.subtype === 'replay-query' ||
-			message.subtype === 'reconstruct-insert'
+			queryOp.subtype === 'replay-query' ||
+			queryOp.subtype === 'reconstruct-insert'
 		) {
-			if (!shouldSyncQuery(message)) {
+			if (!shouldSyncQuery(queryOp)) {
 				return;
 			}
 
 			if (activeTransaction) {
-				activeTransaction.push(message);
+				activeTransaction.push(queryOp);
 			} else {
-				committedQueries.push(message);
+				onCommit([queryOp]);
 			}
 		}
-
-		debouncedFlush();
 	});
-
-	let flushTimeout: number | null = null;
-	function debouncedFlush() {
-		if (null !== flushTimeout) {
-			clearTimeout(flushTimeout);
-		}
-		flushTimeout = setTimeout(() => {
-			flushTimeout = null;
-			const dataToBroadcast = committedQueries;
-			committedQueries = [];
-			onFlush(dataToBroadcast);
-		}, debounceDelay) as any;
-	}
 }
 
-function shouldSyncQuery(query: SQLQueryMetadata) {
-	if (query.query_type === 'SELECT') {
+function shouldSyncQuery(meta: SQLQueryMetadata) {
+	if (meta.query_type === 'SELECT') {
 		return false;
 	}
-	if (query.subtype === 'replay-query') {
+	const queryType = meta.query_type;
+	const tableName = meta.table_name?.toLowerCase();
+
+	if (meta.subtype === 'replay-query') {
+		const query = meta.query.trim();
 		// Don't sync cron updates
 		if (
-			query.query_type === 'UPDATE' &&
-			query.table_name.toLowerCase() === 'wp_options' &&
-			query.query.trim().endsWith("`option_name` = 'cron'")
+			queryType === 'UPDATE' &&
+			tableName === 'wp_options' &&
+			query.endsWith("`option_name` = 'cron'")
 		) {
 			return false;
 		}
 	}
-	if (query.subtype === 'reconstruct-insert') {
+	if (meta.subtype === 'reconstruct-insert') {
 		// Don't sync transients
-		const isTransient =
-			query.table_name.toLowerCase() === 'wp_options' &&
-			typeof query.row.option_name === 'string' &&
-			(query.row.option_name.startsWith('_transient_') ||
-				query.row.option_name.startsWith('_site_transient_'));
-		if (isTransient) {
-			return false;
+		if (tableName === 'wp_options') {
+			const optionName = meta.row.option_name + '';
+			if (
+				optionName.startsWith('_transient_') ||
+				optionName.startsWith('_site_transient_')
+			) {
+				return false;
+			}
```

Comment:
I was wondering if this is the most ideal place to prune and filter queries or if `recordSQLQueries()` inside of `setupPlaygroundSync()` would be a clearer spot. if put there then we could see clearly all queries that pass through the database and provide a guide for adding additional filtering.

in fact, I wonder if there would be a reason to separate synching transactions from logging and/or transforming transactions. for instance, what if this function were called on every query and returns the query. our synching code would follow this journal and record queries of interest, returning the original query (or `null`/`undefined` as a convenience shorthand). suppose we wanted to inject specific transients to setup a test; we could look for these very transients and return a short-circuited DB query. instead of `SELECT * FROM wp_usermeta WHERE meta_key = 'session_tokens'` we return `SELECT 15 as user_id, 'session_tokens' as meta_key, …;`

we could also let this function directly return data, but by returning an SQL query we might have a better chance of guarding the data types exposed to WordPress.

just an idea.

Reply:
I had similar thoughts! I just introduced middlewares to rewrite the site URL, and they can modify the journal as you described. I just moved the pruning phase to a middleware – what do you think?